### DISCUSSION TITLE
ability to right click the heads of top of two programs and click "sync next keyboard interaction"

### DISCUSSION THREAD

**Discussion by RustoMCSpit** [linuxmint/discussions#88] [Ideas]
if i want to sync two yt vids perfectly i could make two windows of firefox and sync them for the next keyboard key and hit space for example.

**Comment by navid-zamani** [comment]
1. That is a way too specific feature to integrate into a general desktop environment. One that’s not even necessary given countless other better solutions.
2. The desktop environment is not the one responsible for these sorts of things. And may not even be able to control them. (Programs are the only ones to decide what they use as input.)

Input is handled by the kernel, X, and Firefox. In that order.
Either, Firefox has to send the keyboard input to multiple windows. (Which the developers would see no point in, so you’d have to write your own patch or extension.)
Or you have to have two different programs (even if just two different Firefox executables that know nothing of each other) and X is altered such that is sends the events to both. (Again, nobody would be interested in doing that, so you’d have to do it yourself.)
Or you run two separate virtual machines with X and Firefox in them and manually handle piping the real input event devices to those two VMs.
Alternatively, you could run two Firefoxes on the same system, but have them connect to different X servers with different keyboards enabled. (xorg.conf allows configuring which keyboards it uses.)

But frankly, I would just not use a stupid browser to do the job of a video player (see also: yt-dlp), and use mplayer’s built-in control interface via named pipes, to have my script send the play command to both processes at the “same” time.  
Though note, that unless you are using two threads on two separate cores that send their commands to two separate mplayer processes running on separate cores too, and you have some kind of timing algorithm to sync those two threads and the mplayer processes too, that still doesn’t guarantee sample-perfect synchronicity. As it only takes 20 microseconds (0.02ms) (or rather half of that) to mess that up.

Alternatively, I’d write my own browser extension or just patch the Firefox code right away if necessary.

There’s probably a trivial solution that I’m forgetting. :)
Sample-perfect timing is a properly hard problem in any case though, even for a professional. Unless you’re writing the video player yourself and know all the delays (e.g. cache misses) and run on a realtime kernel. Then it’s just normally hard. :)

By the way: You may want to use StackExchange for such questions in the future.
